oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/FileChangeMonitor.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.library.util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * File change monitor is a disk file watcher. It keeps to watch the file `last modified timestamp`, after it changed,
```

Comment:
We should consider another scenario if the content of file doesn't change the monitor should not notify the related component.
A possible approach is to compare the binary data of them.

Reply:
> We should consider another scenario if the content of file doesn't change the monitor should not notify the related component.

I expect this is the responsibility of notifier implementation. I just don't want to read the file twice.

Reply:
I prefer to leave it to the monitor since it extracts the input-stream from the target file. And the comparison is a common requirement for all of the notifiers, Do you have some scenario needs to get notification of modification without changing the content? if that exists, it's sane for it.

Reply:
I am not sure whether there is that case, but I prefer to provide an input stream rather than byte[]. 
Let's say in this way, if this class compare the changes, then we need extra memory space for the file contents(whole binary), but if we put the content into implementation, there is no extra memory cause, as you already have username/password configuration in provider config.

I think your idea is only coming from the cert file, as you don't need to read it, right? For ES case, this is a double read or extra memory case.

Reply:
Anyway, I could provide a default `ContentChangedNotifier` implementation, I think that will fit your wants?

Reply:
> I think your idea is only coming from the cert file, as you don't need to read it, right? For ES case, this is a double read or extra memory case.

No. For ES and cert file, double read or extra memory is necessary. Because file contents change leads to a total reconnection. For cert file, all incoming gRPC connections should close as ES does. SSL file notifier also loads the binary into heap to compare the content to avoid unnecessary reconnection.

And I think we should provide an abstract class to implement notifier interface, which can do binary comparison. If anyone needs this feature can extend it instead of implementing notifier.

Reply:
> No. For ES and cert file, double read or extra memory is necessary.

My point is for ES, we don't need extra memory, the user name and password are there, `ContentChangedNotifier` implementation doesn't have to hold it again(one in the ES storage config, the other in the watcher). I know your point to be easier for the watcher, but, this has a cost for ES implementation.

Reply:
All right. Finally, we reach an agreement to provide to implement `ContentChangedNotifier` to provide this feature. Thanks a lot.